Score with cAdvisor metrics (#512)
intel/workload-collocation-agent: examples/kubernetes/monitoring/prometheus/cadvisor.md

## Why

As good as Workload Collocation Agent is, it is a PoC project with small maintainer group and community. For production grade environments something more stable and proven is required, therefore a decision has been made to switch to [cAdvisor](https://github.com/google/cadvisor). Because the way the metrics are gathered and reported differs between the two, it was impossible to take score calculation based on WCA metrics and make it work 1:1 on cAdvisor metrics. Therefore rules has been rewritten to align with new data source.

## How

The description provided is based on prometheus, but the algorithm and part of requirements apply to any system.

### Algorithm

As mentioned in the beginning, the overall goal of the process is to determine if a workload is fit for 2lm node with Intel memory. In general, it answers the question:

*Which will be saturated first on a node with Intel memory dice: memory or any other resource?*

> The answer takes form of ratio how many times faster will the other resource(cpu, memory bandwidth or working set size) be saturated than the memory, for example if score is 1.25 it means that one of the resources will saturate 1.25 times faster than memory. Therefore score interpretation is the following:

Reviewer: missing space before bracket

| Score | Interpretation|
| -----| --------------|

> |0 | workload only allocates memory but doesn't use it or any cpu (perfect candidate) |

Reviewer 2: ```suggestion
| -> 0 | workload which almost do not use any resources apart from allocating memory, perfect match for 2lm |
```

| <1| node filled with such workloads, memory will be 100% utilized when constrained by other resource|
| <3 | not perfect but still worth considering as good candidate for 2LM node|
|>>3 | memory utilization will be low because of saturation on other resources|

As mentioned, algorithm looks at 4 different parameters:

- CPU requirements
- Memory requirements
- Memory bandwidth
- Working set size

Score computation depends on couple of parameters describing applications access to memory:

1. *How much memory does it require?* (the more the better)
2. *How much of this memory is actually used most of the time(what is the WSS(working set size))?* (the smaller the better)

> 3. *What bandwidth can we guarantee for an app in communication between CPU cache and memory?* (the bigger the better)

Reviewer 2: 3 and 4 have the same meaning in my understanding.

Ad 3. what bandwidth is required by app. quarantee is a wrong choice of word

Author: I was trying to refer indirectly to 2 things that we gather: demand data.(rfo|rd) and separately bandwith bytes. I can remove one if required

4. *How often does the app reach to memory outside the CPU cache?*

*NOTE*: In the case of the algorithm, the scope of data is the whole application, not a certain container or a certain pod.

Based on the data gathered answering those questions, algorithm works as follows:

1. Prepare an entry that mocks existence of a PMEM node with some assumed parameters, as it will be required for comparison
2. Measure application instances behavior
3. Aggregate data to represent overall application behavior
4. Assess the most problematic limitation that will be returned as the score

The whole point of the implementation provided as example later on is to prepare all the data required by 5 operations: 3 to get ratio of requirements of cpu, bandwidth and wss in relation to memory, then normalizing it based on the mock pmem node and returnig the max(so the worst possible) score.

### Prerequisites

> List of concepts and requirements to know before implementation explanation.

Author: Point reader to the fact that one should get to know this part prior to deploying

#### Deployed workloads

The algorithm assumes that it is used in production or production-like environment. Input shouldn't be a synthetic workload, but a normal aplication.

#### Workload identification

> The algorithm requires that there will be a way to identify all instances of a workload. E.g. a common label on all pods identifying the workload they belong to(see how "app" label is handled in example).

Reviewer 3: the above issues is outdated, eventually in k8s we have:
- **controller-revision-hash** for both daemonsets and statefulsets e.g. `controller-revision-hash=58cdb6778b`
- **pod-template-hash** for deployments (e.g. `pod-template-hash: 58687784f9`) 

unfortunately using above labels doesn't carry any information about deploy/sts/daemonset name

@wacuuu are there any chances that we can join above "hash" with objects exposed by kube-state-metrics?

> The algorithm requires that there will be a way to identify all instances of a workload. E.g. a common label on all pods identifying the workload they belong to(see how "app" label is handled in example).

Author: Well, controller revision hash is a label on infra container, so it could be added the same way as app label in example: map containers to pod by pod id and inherit all labels from pod container.

> The algorithm requires that there will be a way to identify all instances of a workload. That is there has to be for example a common label on all pods identifying the workload they belong to(see how "app" label is handled in example).

Reviewer 2: ```suggestion
The algorithm requires that there will be a way to identify all instances of a workload. E.g. common label on all pods identifying the workload they belong to (see how "app" label is handled in example).
```

#### Behavior measurement

Shift to cAdvisor has been made, however if one requires, they can use a different software. It has to be able to report memory bandwidth based on resctrl, perf events, specifically offcore_requests.demand_data_rd and offcore_requests.demand_rfo, as well as working set size based on /proc/smaps file. Fortunately cAdvisor can do this all.

*WARNING*: as of today some changes required for full readiness are still awaiting to be merged. For now, [this fork and branch](https://github.com/wacuuu/cadvisor/tree/jwalecki/even-more-magic) contains all needed features.

#### Requirement measurement

Besides measuring application behavior, one must take the requirements into consideration, namely requirement for CPU and memory. This can be easily done by using kube-state-metrics. If this is not possible, one can manually apply the requirements, although it would require to put them for every benchmarked application.

### Implementation

The following part describes how is the example in this repo implemented.

#### Infra description

Example consists of example workloads and monitoring setup.

##### Workloads

> Example workloads are in [this directory](../../workloads). Those are just reference applications used to test the solution. Only thing important for further explanation is the app label with value set to workload name. It is used to identify the workload in rules.

Reviewer 2: ```suggestion
Workloads used by us in internal testing are located in [this directory](../../workloads). Only thing important for notice is the **app** label with value set to workload name. It is used to identify the workload in rules as mentioned in introduction.
```

##### Monitoring

Deployment consist of 3 independent parts: cadvisor, prometheus and kube-state-metrics.

###### kube-state-metrics

Nothing particularly important in this part of deployment. All the details are [here](../kube-state-metrics). It is set up in such a way that it is accessible for prometheus deployed as operator.

###### cadvisor

Deployment as daemonset is defined in [cadvisor directory](../cadvisor). As suggested before, this is using a custom stitched cadvisor docker image, so one must provide it. Two aspects stand out: perf config and binary arguments. As for the perf config, [perf-hmem-cascadelake.json](../cadvisor/perf-hmem-cascadelake.json) is used. Although it contains a bit more than required, it is good for this deployment. On the arguments:

> - `--perf_events_config=/etc/config/perf-hmem-cascadelake.json` is an argument pointing to perf config

Reviewer 2: ```suggestion
- `--perf_events_config=/etc/config/perf-hmem-cascadelake.json` is an argument pointing to perf config, listing needed perf events
```
